### PR TITLE
Parse rework for nested commands

### PR DESCRIPTION
**Pull request by shadowspawn**
# Pull Request

Going wide and trying to unify and improve behaviours at same time as adding support for nested commands.

## Problem

The parsing is a collection of accumulated behaviours, making it difficult to add new features, and in particular nested commands. For example:

- action handlers and executable subcommands have individual historical behaviours
- validation of options done in action handler support rather than independent
- the parsing is scattered between multiple routines

Using the Event Emitter for internal processing is a complication and introduces possible interaction with user defined listeners in middle of internal processing.

Commander started out being flexible and leaving things to caller, but as Commander has become more structured and full featured there is an expectation that some behaviours should be included (e.g. error when subcommand not found, due to typo).

### References

Commander does not support nested commands (without using git-style executables):

- Issue: #1 #63 #655 #764
- Related PR: #245 #1024

Allow a non-executable to be set as the default command: #742

Implicit help command: #205 #256 #350 #892
but not when an argument: #1153

Implicit help not working for alias: #948 

Built-in error when unrecognised command: #432 #1088

Help displayed for missing command: #1088 

Built-in error for missing arguments: #995 

Asterisk handler getting called for executable subcommands: #809

## Solution

Rework parsing, big time:

- do not use events for internal command processing (still used for option processing)
- explicitly handle just one of action vs executable
- parse command independent of whether action handler attached
- recursive parsing, for nested command support
- recursive parsing, with intention top level command is less special
- treat action/executable subcommands more symmetrically
- add error and/or help for missing and unknown commands (for all subcommands, not just executable)

### Added

- allow a non-executable to be set as the default command (#742 #1149)
- support nested sub-commands with action-handlers (#1 #764 #1149)
- automatic implicit help command for all subcommands (not just executables) (#1149)
- customise implicit help command with `.addHelpCommand()` (#1149)
- _Breaking_ display error message for unknown subcommand (#432 #1088 #1149)
- _Breaking_ display help for missing subcommand (#1088 #1149)

### Changed

- _Breaking_  display error if required argument for command is missing (#995 #1149)

### Fixed

- only add implicit help command when it will not conflict with other arguments (#1153 #1149)
- do not emit `command:*` for executable subcommands (#809 #1149)
- implicit help command works with aliases (#948 #1149)
- options are validated whether or not there is an action handler (#1149)

### Removed

- `parseExpectedArgs` is now private (#1149)

### Migration tips

If you use `.on('command:*')` or more complicated tests to detect an unrecognised subcommand, you may be able to delete the code and rely on the default behaviour.

If you use `program.args` or more complicated tests to detect a missing subcommand, you may be able to delete the code and rely on the default behaviour.

If you use `.command('*')` to add a default command, you may be be able to use `isDefault:true` with a normal command instead.

### CODE REVIEW

**Comment by byCedric** [comment]
Awesome work! 🚀

**Comment by shadowspawn** [comment]
I am thinking of moving onto `develop` after this PR?

The other things I am considering for 5.0 are largely non-breaking:
- change all setup errors by author to `throw` rather than error+exit (about half already do), and add tips to the infamous accidental executable situation
- move `Command` from prototype to class (technically non-breaking, but core change)
- switch from `var` to `const` and `let` (non-breaking)
- make `.helpInformation()` public and rework README section on help (non-breaking)
- consider more options to .parse for electron et al (technically non-breaking)

File: index.js [hunk 1]
```diff
@@ -124,19 +122,28 @@ exports.CommanderError = CommanderError;
...
   this.options = [];
-  this._execs = new Set();
   this._allowUnknownOption = false;
   this._args = [];
   this._name = name || '';
   this._optionValues = {};
   this._storeOptionsAsProperties = true; // backwards compatible by default
   this._passCommandToAction = true; // backwards compatible by default
   this._actionResults = [];
+  this._actionHandler = undefined;
```
Comment (abetomo) [inline index.js]: I'm sorry if I sound picky.
Here we 'define' it.So I think it's not 'undefined'.
I think 'null' is better, but how about?

Comment (shadowspawn) [reply index.js]: I did feel "defining" it to be "undefined" was a bit odd. I was really just adding it to be clear it was a property. I'll do a little reading on null vs undefined conventions. Good question.

Comment (shadowspawn) [reply index.js]: Changed all to null except the boolean that is lazy initialised where undefined felt better (and added a comment for that one).

**Comment by shadowspawn** [comment]
Woo hoo! 🎉 
Thanks @abetomo , it was a big and complex Pull Request to review.

I'll merge it in soon and start work on preparing CHANGELOG / Release Notes for a prerelease.

**Comment by shadowspawn** [comment]
I gave the nested command support a try-out emulating the Cordova CLI. I implemented one nested subcommand using inline code, and one using `.addCommand` and a module:
https://github.com/shadowspawn/commander-does-cordova.git

**Pull request merged**